File: src/node/magma/divide/MutableDivideState.ts
/*import magma.Tuple;*/
/*import java.util.ArrayList;*/
/*import java.util.Collection;*/
/*import java.util.Optional;*/
/*import java.util.stream.Stream;*/
export class MutableDivideState implements DivideState {
	/*private final Collection<String> segments = new ArrayList<>()*/;
	/*private final CharSequence input*/;
	/*private int index = 0*/;
	/*private StringBuilder buffer = new StringBuilder()*/;
	/*private int depth = 0*/;
	/*public MutableDivideState(final CharSequence input) {
        this.input = input;
    }*/
	/*@Override
    public DivideState advance() {
        this.segments.add(this.buffer.toString());
        this.buffer = new StringBuilder();
        return this;
    }*/
	/*@Override
    public DivideState append(final char c) {
        this.buffer.append(c);
        return this;
    }*/
	/*@Override
    public Stream<String> stream() {
        return this.segments.stream();
    }*/
	/*@Override
    public boolean isLevel() {
        return 0 == this.depth;
    }*/
	/*@Override
    public DivideState enter() {
        this.depth++;
        return this;
    }*/
	/*@Override
    public DivideState exit() {
        this.depth--;
        return this;
    }*/
	/*@Override
    public Optional<Tuple<DivideState, Character>> pop() {
        if (this.index >= this.input.length()) return Optional.empty();
        final var c = this.input.charAt(this.index);
        this.index++;
        return Optional.of(new Tuple<>(this, c));
    }*/
	/*@Override
    public Optional<Tuple<DivideState, Character>> popAndAppendToTuple() {
        return this.pop().map(tuple -> {
            final var appended = tuple.left().append(tuple.right());
            return new Tuple<>(appended, tuple.right());
        });
    }*/
	/*@Override
    public Optional<DivideState> popAndAppendToOption() {
        return this.popAndAppendToTuple().map(Tuple::left);
    }*/
	/*@Override
    public boolean isShallow() {
        return 1 == this.depth;
    }*/
}/**/

File: src/java/magma/divide/MutableDivideState.java
package magma.divide;

import magma.Tuple;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import java.util.stream.Stream;

public class MutableDivideState implements DivideState {
	private final String input;
	private final Collection<String> segments;
	private final StringBuilder buffer;
	private int index;
	private int braceCount;

	public MutableDivideState(final String input) {
		this.input = input;
		this.index = 0;
		this.braceCount = 0;
		this.segments = new ArrayList<>();
		this.buffer = new StringBuilder();
	}

	@Override
	public final Optional<Tuple<DivideState, Character>> pop() {
		final var length = this.input.length();
		if (this.index >= length) return Optional.empty();
		final char c = this.input.charAt(this.index);

		this.index = this.index + 1;
		return Optional.of(new Tuple<>(this, c));
	}

	@Override
	public final boolean isLevel() {
		return 0 == this.braceCount;
	}

	@Override
	public final DivideState exit() {
		this.braceCount = this.braceCount - 1;
		return this;
	}

	@Override
	public final DivideState enter() {
		this.braceCount = this.braceCount + 1;
		return this;
	}

	@Override
	public final DivideState advance() {
		this.segments.add(this.buffer.toString());
		this.buffer.setLength(0);
		return this;
	}

	@Override
	public DivideState append(final char c) {
		this.buffer.append(c);
		return this;
	}

	@Override
	public final Stream<String> stream() {
		return this.segments.stream();
	}
}
